BREAKING CHANGE(core): Make throwForStatus an opt-out, also for code/function operations

FokkeZB opened the pull request:
> **NOTE:** Replaces #193

## Description

Make throwForStatus an opt-out, also for code-mode/function operations.

### Why?

1. We up its share in v3 errors from 10 closer to the 60% on v2 so that:
    1. We have more control over the (default) error handling, which recently has been updated to extract messages and codes from error responses like we already did for v1/2.
    2. We lower the 20% "errors" that contain JSON, HTML or Node Runtime errors (pointing at no/poor error handling) on v3 closer to the 1% on v2.
2. We force developers to use `afterResponse` for custom error handling if `throwForStatus` throws when it shouldn't (or doesn't - correctly - when it should).

This requires the visual builder to also support defining a (single) `afterResponse` middleware, which is what we've shipped via https://github.com/zapier/zapier/pull/39765 for staff-only as part of the 2020 (remote) engineering retreat, and can make public via https://github.com/zapier/zapier/pull/39899.

## Before

### Shorthand/form-mode Requests

We always run `throwForStatus`, following any `afterResponse`, meaning:

1. You cannot use shorthand/form-mode for API requests that may return a status > 300 **when there is no error**.
2. You cannot use shorthand/form-mode for API requests that may return a status <= 300 **when there is an error**, because `throwForStatus` will be called but not throw, after which we'll try to parse `response.content` as a successful result, which will either throw an unhelpful error if we fail to do so or break the Zap as fields will be different.

In CLI apps and after https://github.com/zapier/zapier/pull/39899 you can handle this via `afterResponse`, but for 1) this is hacky:

```node
SHOULD_NOT_THROW = 444;
SHOULD_THROW = 222;

if (response.status === SHOULD_NOT_THROW) {
  response.status = 200;
}

if (response.status === SHOULD_THROW) {
  throw z.errors.Error(response.json.errorMessage);

  // or hacky:
  response.status = 500;
}

return response;
```

### Function/Code-Mode Request

`throwForStatus` is an (easy to forget) opt-in:

```node
return z.request({ ... })
  .then(response => {
    response.throwForStatus();
    return response.json;
  }); 
```

## After

### Function/Code-Mode Requests

No more need to (remember to) call `throwForStatus`:

```node
return z.request({ ... })
  .then(response => response.json);
```

Ability to opt-out of `throwForStatus` for specific requests:

```node
return z.request({
  ...
  skipThrowForStatus: true
})
  .then(response => {
    requestSpecificErrorHandling(response);
    return response.json;
  });
```

### After Response Middleware

Ability to opt-out of `throwForStatus` for specific responses:

```node
SHOULD_NOT_THROW = 444;

if (response.status === SHOULD_NOT_THROW) {
    response.skipThrowForStatus = true;
}
return response;
```

## Related

**Initial Spec:** https://github.com/zapier/sandbox-integration-errors/issues/7
**Updated Spec:** https://docs.google.com/document/d/1j1O2_88dirRvciAEO05axE_B1J2tCfKM8qsGRshDepc/edit#heading=h.fowdhe4cl4hl

## To Do

- [x] Check how this impacts `legacy-scripting-runner`

    These are code-mode requests and will be detected as such by the migration in https://github.com/zapier/zapier/pull/39898 which will make sure they don't break. However, this PR also adds extra safeguards to `legacyScripting.run` in case they want to opt-in new code-mode/function requests while not breaking legacy actions.

- [x] Add internal-only Middleware tab to visual builder: https://github.com/zapier/zapier/pull/39765
- [ ] Run a migration to run on existing visual builder apps to not break them: https://github.com/zapier/zapier/pull/39898
- [ ] Update templates in `zapier/zapier` to no longer include `throwForStatus` for code-mode: https://github.com/zapier/zapier/pull/39900
- [ ] Make the Middleware tab public: https://github.com/zapier/zapier/pull/39899

FokkeZB commented on example-apps/babel/src/authentication.js:
```diff
@@ -5,16 +5,6 @@ const test = async (z /*, bundle */) => {
-
-  // This method can return any truthy value to indicate the credentials are valid.
-  // Raise an error to show
-  if (response.status === 401) {
-    throw new z.errors.Error(
```
Examples have been updates to no longer test for error response status as `throwForStatus` is now always called.

FokkeZB commented on example-apps/babel/src/test/index.js:
```diff
@@ -20,20 +20,14 @@ describe('My Test', () => {
-        'The username and/or password you supplied is incorrect'
-      );
-    }
+    return appTester(App.authentication.test, bundle).should.be.rejected();
```
These now reject with a `z.errors.ResponseError` instead of a custom error.

FokkeZB commented on packages/cli/README-source.md:
```diff
@@ -994,6 +994,10 @@ Middleware functions can be asynchronous - just return a promise from the middle
 
 The second argument for middleware is the `z` object, but it does *not* include `z.request()` as using that would easily create infinite loops.
 
+#### Error Response Handling
```
Documents the new behavior.

FokkeZB commented on packages/cli/README-source.md:
```diff
@@ -1012,6 +1016,7 @@ Shorthand requests and manual `z.request([url], options)` calls support the foll
 * `agent`: Node.js `http.Agent` instance, allows custom proxy, certificate etc. Default is `null`.
 * `timeout`: request / response timeout in ms. Set to `0` to disable (OS limit still applies), timeout reset on `redirect`. Default is `0` (disabled).
 * `size`: maximum response body size in bytes. Set to `0` to disable. Default is `0` (disabled).
+* `skipThrowForStatus`: don't call `response.throwForStatus()` before returning it.
```
Documents the new flag.

FokkeZB commented on packages/cli/README-source.md:
```diff
@@ -1213,11 +1219,14 @@ formulate valid requests. Thus, it is a good idea to write apps defensively and
-various kinds of errors occur.
+Zapier provides a couple tools to help with error handling. First is the
+`afterResponse` middleware ([docs](#using-http-middleware)), which provides a hook for
+processing all responses from HTTP calls. Second is `response.throwForStatus()`
```
Updated to include the role of `throwForStatus` in error handling.

FokkeZB commented on packages/schema/lib/schemas/RequestSchema.js:
```diff
@@ -80,6 +80,12 @@ module.exports = makeSchema(
           'A function to customize how to serialize a value for curlies `{{var}}` in the request object. By default, when this is unspecified, the request client only replaces curlies where variables are strings, and would throw an error for non-strings. The function should accepts a single argument as the value to be serialized and return the string representation of the argument.',
         $ref: FunctionSchema.id
+      },
+      skipThrowRequest: {
```
Updated `RequestSchema` to include new flag.

eliangcs replied:
Typo:

```suggestion
      skipThrowForStatus: {
```

FokkeZB commented on packages/schema/lib/schemas/AuthenticationOAuth2ConfigSchema.js:
```diff
@@ -33,7 +33,7 @@ module.exports = makeSchema(
       autoRefresh: {
         description:
-          'Should Zapier include a pre-built afterResponse middleware that invokes `refreshAccessToken` when we receive a 401 response?',
+          'Should Zapier invoke `refreshAccessToken` when we receive an error for a 401 response?',
```
Updated description for new behavior without `throwForStaleAuth`.

FokkeZB commented on example-apps/middleware/index.js:
```diff
@@ -15,15 +15,17 @@ const addSortingParams = (request /*, z */) => {
-    );
+const handleErrors = (response, z) => {
+  // Prevent `throwForStatus` from throwing for a certain status.
+  if (response.status === 456) {
+    response.skipThrowForStatus = true;
+  }
```
Updated middleware example app to demonstrate new flag, identical to documentation middleware snippet.

FokkeZB commented on example-apps/session-auth/index.js:
```diff
@@ -30,7 +20,7 @@ const App = {
 
   beforeRequest: [includeSessionKeyHeader],
 
-  afterResponse: [sessionRefreshIf401],
```
There was never a need for this as we already used to always call `throwForStaleAuth` for session auth.

FokkeZB commented on packages/cli/scaffold/create.template.js:
```diff
@@ -9,7 +9,6 @@ const perform = async (z, bundle) => {
       name: bundle.inputData.name
     }
   });
-  response.throwForStatus();
```
Updated scaffold templates to no longer manually call `throwForStatus`.

FokkeZB commented on packages/cli/snippets/middleware.js:
```diff
@@ -3,22 +3,23 @@ const addHeader = (request, z, bundle) => {
-      response.status
-    );
+const handleErrors = (response, z) => {
+  // Prevent `throwForStatus` from throwing for a certain status.
+  if (response.status === 456) {
+    response.skipThrowForStatus = true;
```
Updated middleware snippet to demonstrate new flag.

FokkeZB commented on packages/cli/src/tests/utils/astFixtures.js:
```diff
@@ -58,6 +58,10 @@ const beforeRequest = (request, z, bundle) => {
 };
 
 const afterResponse = (response, z, bundle) => {
+  if (response.request.isShortHand === false) {
+    response.skipThrowForStatus = true;
+  }
+
   return z.legacyScripting.afterResponse(response, z, bundle);
```
Updated legacy `afterResponse` snippets to be what https://github.com/zapier/zapier/pull/39898 changes them to.

FokkeZB commented on packages/core/index.d.ts:
```diff
@@ -90,13 +90,15 @@ export interface HttpRequestOptions {
   size?: number;
   timeout?: number;
   url?: string;
+  skipThrowForStatus: boolean;
```
Typing for the new flag.

FokkeZB commented on packages/core/src/constants.js:
```diff
@@ -18,7 +18,7 @@ const RENDER_ONLY_METHODS = [
 ];
 
-const REQUEST_OBJECT_SHORTHAND_OPTIONS = { replace: true };
+const REQUEST_OBJECT_SHORTHAND_OPTIONS = { isShorthand: true, replace: true };
```
Required for https://github.com/zapier/zapier/pull/39898

FokkeZB commented on packages/core/src/execute.js:
```diff
@@ -9,31 +9,27 @@ const createJSONtool = require('./tools/create-json-tool');
   options = _.extend({}, options, constants.REQUEST_OBJECT_SHORTHAND_OPTIONS);
-  return input.z
-    .request(options)
-    .then(throwForStatus)
```
No more need to manually call `throwForStatus` after form-mode/short-hand requests.

FokkeZB commented on packages/core/src/http-middlewares/after/prepare-response.js:
```diff
@@ -8,7 +8,8 @@ const prepareRawResponse = (resp, request) => {
   const extendedResp = {
-    request: request
+    request: request,
+    skipThrowForStatus: request.skipThrowForStatus
```
Allows setting the new flag when initiating a request instead of only from `afterResponse`.

FokkeZB commented on packages/core/src/http-middlewares/after/throw-for-status.js:
```diff
@@ -2,12 +2,12 @@
-const throwForStatus = resp => {
-  if (resp.status > 300) {
-    throw new errors.ResponseError(resp);
+const throwForStatus = response => {
+  if (!response.skipThrowForStatus && response.status > 300) {
```
Don't throw when the new flag is used.

FokkeZB commented on packages/core/src/http-middlewares/before/prepare-request.js:
```diff
@@ -105,14 +105,15 @@ const finalRequest = req => {
       body: false
     },
     replace: true, // always replace curlies
+    skipThrowForStatus: false,
```
Although `undefined` is falsy too, make sure the new flag is there on the request

FokkeZB commented on packages/core/src/tools/create-app-request-client.js:
```diff
@@ -51,19 +51,13 @@ const createAppRequestClient = (input, options) => {
+  const beforeAppAfterResponse = [prepareResponse, logResponse];
+  const afterAppAfterResponse = [throwForStatus];
 
-  if (app.authentication) {
-    if (
-      app.authentication.type === 'session' ||
-      (app.authentication.type === 'oauth2' &&
-        _.get(app, 'authentication.oauth2Config.autoRefresh'))
-    ) {
-      httpOriginalAfters.push(throwForStaleAuth);
-    }
-  }
```
No more need for `throwForStaleAuth` as we now always call `throwForStatus` which also throws on 401 and we already reraise that as a `RefreshAuthError` on server/python-side for both cases.

FokkeZB commented on packages/core/src/tools/create-app-request-client.js:
```diff
@@ -51,19 +51,13 @@ const createAppRequestClient = (input, options) => {
 
-  const httpOriginalAfters = [prepareResponse, logResponse];
+  const beforeAppAfterResponse = [prepareResponse, logResponse];
+  const afterAppAfterResponse = [throwForStatus];
```
This is where we now always run `throwForStatus` after `afterResponse`.

FokkeZB commented on packages/core/test/create-request-client.js:
```diff
@@ -270,12 +271,27 @@ describe('request client', () => {
   });
 
-  it('should run any afterResponse functions', done => {
+  it('should default to run throwForStatus', () => {
```
Tests for the new flag.

FokkeZB commented on packages/core/test/errors.js:
```diff
@@ -27,11 +25,16 @@ describe('errors', () => {
 
   describe('ResponseError', () => {
     it('should stringify arguments', async () => {
```
Simplified test

FokkeZB commented on packages/core/test/userapp/index.js:
```diff
@@ -500,33 +500,9 @@ const addRequestHeader = (request, z, bundle) => {
-  changes the status code if there's an error. Something like this
-  could be useful for APIs that always return 200 even on errors.
- */
-const changeStatusOnErrorResponses = response => {
```
Wasn't used anywhere and was a bad example.

FokkeZB commented on packages/legacy-scripting-runner/index.js:
```diff
@@ -672,6 +672,7 @@ const legacyScriptingRunner = (Zap, zcli, input) => {
       request.headers = cleanHeaders(request.headers);
       request.allowGetBody = true;
       request.serializeValueForCurlies = serializeValueForCurlies;
+      request.skipThrowForStatus = true;
```
This makes sure that even if a user deletes what https://github.com/zapier/zapier/pull/39898 adds to `afterResponse` for legacy (visual builder) apps, we are still in control of when we call `throwForStatus`.

FokkeZB commented on packages/legacy-scripting-runner/middleware-factory.js:
```diff
@@ -203,18 +203,21 @@ const createBeforeRequest = app => {
 
-  if (authType === 'session') {
-    afterResponse = sessionAuthCheckResponse;
+  if (authType === 'session' || (authType === 'oauth2' && autoRefresh)) {
+    afterResponse = throwForStaleAuth;
```
For some reason the legacy runner relied on `throwForStaleAuth` for oauth2 only, but now it has to do so for both session and oauth2.

FokkeZB commented on packages/legacy-scripting-runner/test/example-app/index.js:
```diff
@@ -1171,6 +1183,19 @@ const App = {
       source: "return z.legacyScripting.run(bundle, 'hydrate.file');"
     }
   },
+  // This breaks `applyBeforeMiddleware` which it looks like gets the original definition without `findSourceRequireFunctions` applied?!
+  // beforeRequest: [
```
I found a bug but wasn't able or required to fix it now.

eliangcs replied:
Nice finding! Looks like it's an issue with the test code. Those failing tests don't compile the app definition but they should. I'll push a fix once this is merged.

eliangcs commented on packages/schema/lib/schemas/RequestSchema.js:
```diff
@@ -80,6 +80,12 @@ module.exports = makeSchema(
+      },
+      skipThrowRequest: {
+        description:
+          'Should throwForStatus be calld before returning the response?',
```
Ditto, we can be more specific by saying:

> If true, throw an exception for response status >= 400 automatically before returning the response. Defaults to `false`.

FokkeZB replied:
Done, but using `> 300` for now.

xavdid commented:
Like Chang-hung mentioned, some of our eng retreat work concerned these example apps. We'll be removing some of them long term. Before we release, we'll probably want you to take a second look at the example auth output to make sure everything has been in sync. I made the new templates based on the files in https://github.com/zapier/zapier-platform/tree/master/examples/auth

FokkeZB commented on example-apps/onedrive/authentication.js:
```diff
@@ -49,22 +49,11 @@ const getAccessToken = (z, bundle) => {
+    access_token: response.json.access_token,
+    refresh_token: response.json.refresh_token,
+    id_token: response.json.id_token
+  }))
```
As it's done already, I'll leave it in.

FokkeZB commented on packages/cli/README-source.md:
```diff
@@ -994,6 +994,10 @@ Middleware functions can be asynchronous - just return a promise from the middle
+#### Error Response Handling
+
+Since v10, we call `response.throwForStatus()` before we return a response. You can prevent this by setting `skipThrowForStatus` on the request or response object. You can do this in `afterResponse` middleware if the API uses a status code > 300 that should not be treated as an error.
+
```
Added!

FokkeZB commented on packages/cli/README-source.md:
```diff
@@ -1213,11 +1219,14 @@ formulate valid requests. Thus, it is a good idea to write apps defensively and
+`afterResponse` middleware ([docs](#using-http-middleware)), which provides a hook for
+processing all responses from HTTP calls. Second is `response.throwForStatus()`
+([docs](#http-response-object)), which throws an error if the response status indicates
+an error. We automatically call this method before returning the response, unless
```
Done. We currently throw for > 300 so I've used that, but once you approved https://github.com/zapier/zapier-platform/pull/192 we can ship that in the same breaking and I'll make sure to update the docs on whatever PR merges last.

FokkeZB commented on packages/core/src/tools/create-app-request-client.js:
```diff
@@ -51,19 +51,13 @@ const createAppRequestClient = (input, options) => {
+  const httpAfters = beforeAppAfterResponse.concat(
+    ensureArray(app.afterResponse),
+    afterAppAfterResponse
+  );
```
Much better. Done!

FokkeZB commented:
@eliangcs thanks for the thorough review! I've addressed all feedback.

I'll merge once lights are green and then update #192 to update `> 300` references this introduces to `400 <= status < 600`.

@xavdid I'll then also look at #206.

FokkeZB commented:
For documentation-sake, I've also updated this PRs description with before/after examples.
